### PR TITLE
Add SushiswapGoblin

### PR DESCRIPTION
**Pull request by mikephul**
- Migrate to buidler (truffle5)
- Add 2 goblins, SushiswapGoblin and SushiswapPool12Goblin (SUSHI-ETH pool) to support Sushiswap.
- Add StrategyAddTwoSidesOptimal to support supplying ERC20 token along with ETH.

### CODE REVIEW

File: contracts/5/SushiswapGoblin.sol [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    Strategy public addStrat;
+    Strategy public liqStrat;
+    uint256 public reinvestBountyBps;
+
+    constructor(
+        address _operator,
+        IMasterChef _masterChef,
+        IUniswapV2Router02 _router,
+        uint256 _pid,
+        address _sushi,
```
Comment (GriphookETH) [inline contracts/5/SushiswapGoblin.sol:50]: Can we derive this from MasterChef?

File: contracts/5/SushiswapGoblin.sol [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+        if (totalShare == 0) return balance; // When there's no share, 1 share = 1 balance.
+        (uint256 totalBalance, ) = masterChef.userInfo(pid, address(this));
+        return balance.mul(totalShare).div(totalBalance);
+    }
+
+    /// @dev Re-invest whatever this worker has earned back to staked LP tokens.
+    function reinvest() public onlyEOA nonReentrant {
+        // 1. Withdraw all the rewards.
+        (uint256 totalBalance, ) = masterChef.userInfo(pid, address(this));
+        masterChef.withdraw(pid, totalBalance);
```
Comment (GriphookETH) [inline contracts/5/SushiswapGoblin.sol:111]: We can withdraw 0 to withdraw all rewards without touching LP token

File: contracts/5/SushiswapGoblin.sol [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    /// @dev Return the amount of ETH to receive if we are to liquidate the given position.
+    /// @param id The position ID to perform health check.
+    function health(uint256 id) external view returns (uint256) {
+        // 1. Get the position's LP balance and LP total supply.
+        uint256 lpBalance = shareToBalance(shares[id]);
+        uint256 lpSupply = lpToken.totalSupply(); // Ignore pending mintFee as it is insignificant
+        // 2. Get the pool's total supply of WETH and farming token.
+        uint256 totalWETH = weth.balanceOf(address(lpToken));
```
Comment (GriphookETH) [inline contracts/5/SushiswapGoblin.sol:171]: Please update to the latest goblin version (use lpToken.reserves())

File: contracts/5/SushiswapPool12Goblin.sol [hunk 4]
```diff
@@ -0,0 +1,246 @@
+pragma solidity 0.5.16;
```
Comment (GriphookETH) [inline contracts/5/SushiswapPool12Goblin.sol:1]: ditto here for all changes in SushiSwapGobin.

File: contracts/5/SushiswapPool12Goblin.sol [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+    event RemoveShare(uint256 indexed id, uint256 share);
+    event Liquidate(uint256 indexed id, uint256 wad);
+
+    /// @notice Immutable variables
+    IMasterChef public masterChef;
+    IUniswapV2Factory public factory;
+    IUniswapV2Router02 public router;
+    IUniswapV2Pair public lpToken;
+    address public weth;
+    address public fToken;
```
Comment (GriphookETH) [inline contracts/5/SushiswapPool12Goblin.sol:32]: Why do we have fToken? It must be sushi right?

**Review by GriphookETH — CHANGES_REQUESTED** [review]

File: contracts/5/SushiswapGoblin.sol (hunk 1 above)
Comment (mikephul) [reply contracts/5/SushiswapGoblin.sol:50]: Can. Use `sushi = address(masterChef.sushi())` instead.

File: contracts/5/SushiswapGoblin.sol (hunk 2 above)
Comment (mikephul) [reply contracts/5/SushiswapGoblin.sol:111]: Fixed

File: contracts/5/SushiswapPool12Goblin.sol (hunk 5 above)
Comment (mikephul) [reply contracts/5/SushiswapPool12Goblin.sol:32]: Change `fToken` to `sushi`

File: contracts/5/SushiswapGoblin.sol (hunk 3 above)
Comment (mikephul) [reply contracts/5/SushiswapGoblin.sol:171]: Updated to `lpToken.reserves()`

File: contracts/5/StrategyAddTwoSidesOptimal.sol [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+contract StrategyAddTwoSidesOptimal is Ownable, ReentrancyGuard, Strategy {
+    using SafeToken for address;
+    using SafeMath for uint256;
+
+    IUniswapV2Factory public factory;
+    IUniswapV2Router02 public router;
+    address public weth;
+
+    /// @dev Create a new liquidate strategy instance.
```
[the remark this answers is not in the document]
Comment (mikephul) [reply contracts/5/StrategyAddTwoSidesOptimal.sol:20]: Done

File: contracts/5/SushiswapPool12Goblin.sol (hunk 4 above)
Comment (mikephul) [reply contracts/5/SushiswapPool12Goblin.sol:1]: Added comment

File: contracts/5/SushiswapPool12Goblin.sol [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        Strategy _addStrat,
+        Strategy _liqStrat,
+        uint256 _reinvestBountyBps
+    ) public {
+        operator = _operator;
+        weth = _router.WETH();
+        masterChef = _masterChef;
+        router = _router;
+        factory = IUniswapV2Factory(_router.factory());        
+        pid = 12;
```
Comment (GriphookETH) [inline contracts/5/SushiswapPool12Goblin.sol:59]: let's make this a constant at the top. `uint256 public constant pid = 12`

**Review by GriphookETH — CHANGES_REQUESTED** [review]
one small comment. otherwise LGTM

File: contracts/5/SushiswapPool12Goblin.sol (hunk 7 above)
Comment (mikephul) [reply contracts/5/SushiswapPool12Goblin.sol]: done

**Pull request merged**